### PR TITLE
Windows/online

### PR DESCRIPTION
**Pull request by wopeizl**

### CODE REVIEW

File: cmake/external/glog.cmake [hunk 1]
```diff
@@ -46,7 +46,11 @@ ExternalProject_Add(
     CMAKE_ARGS      -DCMAKE_CXX_COMPILER=${CMAKE_CXX_COMPILER}
                     -DCMAKE_C_COMPILER=${CMAKE_C_COMPILER}
                     -DCMAKE_CXX_FLAGS=${CMAKE_CXX_FLAGS}
+                    -DCMAKE_CXX_FLAGS_RELEASE=${CMAKE_CXX_FLAGS_RELEASE}
+                    -DCMAKE_CXX_FLAGS_DEBUG=${CMAKE_CXX_FLAGS_DEBUG}
                     -DCMAKE_C_FLAGS=${CMAKE_C_FLAGS}
+                    -DCMAKE_C_FLAGS_DEBUG=${DCMAKE_C_FLAGS_DEBUG}
```
Comment (reyoung) [inline cmake/external/glog.cmake]: Typo here?

File: python/setup.py.in [hunk 2]
```diff
@@ -175,36 +180,47 @@ if '${CMAKE_BUILD_TYPE}' == 'Release':
...
+                command = "install_name_tool -id \"@loader_path/../paddle/libs/\" ${PADDLE_BINARY_DIR}/python/py_paddle/_swig_paddle" + ext_name
+            else:
+                command = "patchelf --set-rpath '$ORIGIN/../paddle/libs/' ${PADDLE_BINARY_DIR}/python/py_paddle/_swig_paddle" + ext_name
+            if os.system(command) != 0:
+                raise Exception("patch _swig_paddle.so failed, command: %s" % command)
+
+ext_modules = [Extension('_foo', ['stub.cc'])]
+if os.name == 'nt':
+    # fix the path separator under windows
+    fix_package_dir = {}
```
Comment (reyoung) [inline python/setup.py.in:224]: Perhaps we can change all '.so' ==> '.dll' here?

File: cmake/external/boost.cmake [hunk 3]
```diff
@@ -28,34 +28,47 @@ if((NOT DEFINED BOOST_TAR) OR (NOT DEFINED BOOST_URL))
...
-    && tar zxf ${BOOST_TAR}.tar.gz
-    DOWNLOAD_NO_PROGRESS  1
-    PREFIX                ${BOOST_SOURCES_DIR}
-    CONFIGURE_COMMAND     ""
-    BUILD_COMMAND         ""
-    INSTALL_COMMAND       ""
-    UPDATE_COMMAND        ""
-)
-endif(NOT WIN32)
+if (WIN32)
```
Comment (dzhwinter) [inline cmake/external/boost.cmake]: how about just define different DOWNLOAD_COMMAND ?

File: cmake/external/gflags.cmake [hunk 4]
```diff
@@ -48,8 +53,8 @@ ExternalProject_Add(
 IF(WIN32)
   IF(NOT EXISTS "${GFLAGS_INSTALL_DIR}/lib/libgflags.lib")
     add_custom_command(TARGET extern_gflags POST_BUILD
-    COMMAND cmake -E rename ${GFLAGS_INSTALL_DIR}/lib/gflags_static.lib ${GFLAGS_INSTALL_DIR}/lib/libgflags.lib
-  )
+            COMMAND cmake -E copy ${GFLAGS_INSTALL_DIR}/lib/gflags_static.lib ${GFLAGS_INSTALL_DIR}/lib/libgflags.lib
```
Comment (dzhwinter) [inline cmake/external/gflags.cmake:56]: I use rename on purpose, I thought only leave one copy is better.

File: cmake/generic.cmake [hunk 5]
```diff
@@ -288,6 +292,50 @@ function(cc_library TARGET_NAME)
...
+# after:
+#    1. lib /out:dummy_lib_1.lib a.lib b.lib
+#    2. lib /out:dummy_lib_2.lib c.lib d.lib
+#    1. lib /out:target.lib dummy_lib_1.lib dummy_lib_2.lib
+function(sep_library TARGET_NAME)
+  set(options STATIC static SHARED shared)
+  set(oneValueArgs "")
+  set(multiValueArgs SRCS DEPS)
+  set(${TARGET_NAME}_dummy_flag "")
+  if(${sep_library_STATIC})
```
Comment (dzhwinter) [inline cmake/generic.cmake]: split and merge should only support static library, throw an error if this function used to build shared library.

File: paddle/fluid/framework/ir/attention_lstm_fuse_pass.cc [hunk 6]
```diff
@@ -211,12 +211,12 @@ void PrepareLSTMWeight(const LoDTensor& W_forget_w0,
   VLOG(3) << "LSTMWeight resized to " << out->dims();
 
   float* out_data = out->mutable_data<float>(platform::CPUPlace());
-  std::array<const float*, 4> tensors(
-      {{W_forget_w0.data<float>(), W_input_w0.data<float>(),
-        W_output_w0.data<float>(), W_cell_w0.data<float>()}});
-  std::array<const float*, 4> tensors1(
-      {{W_forget_w1.data<float>(), W_input_w1.data<float>(),
-        W_output_w1.data<float>(), W_cell_w1.data<float>()}});
+  std::array<const float*, 4> tensors =
```
Comment (dzhwinter) [inline paddle/fluid/framework/ir/attention_lstm_fuse_pass.cc]: array use = will generate compiler error in apple clang.

File: paddle/fluid/inference/api/api_impl.cc [hunk 7]
```diff
@@ -75,6 +75,10 @@ bool NativePaddlePredictor::Init(
   }
 #endif
 
+  // windows has no support for openblas multi-thread
```
Comment (dzhwinter) [inline paddle/fluid/inference/api/api_impl.cc]: this one should be set in init.cc

**Review by dzhwinter — COMMENTED** [review]
一部分code已经在我的PR里了，merge一下develop分支吧

File: cmake/external/glog.cmake (hunk 1 above)
Comment (wopeizl) [reply cmake/external/glog.cmake]: yes

File: cmake/external/boost.cmake (hunk 3 above)
Comment (wopeizl) [reply cmake/external/boost.cmake]: ok

File: cmake/external/gflags.cmake (hunk 4 above)
Comment (wopeizl) [reply cmake/external/gflags.cmake:56]: It seems use the rename can trigger issue under some cases especially when you try to rebuild, not found the root cause yet, I would prefer to keep it as is and fix it later.

File: paddle/fluid/framework/ir/attention_lstm_fuse_pass.cc (hunk 6 above)
Comment (wopeizl) [reply paddle/fluid/framework/ir/attention_lstm_fuse_pass.cc]: I didn't notice that yet, any advise to fix that?

File: paddle/fluid/inference/api/api_impl.cc (hunk 7 above)
Comment (wopeizl) [reply paddle/fluid/inference/api/api_impl.cc]: Better idea.

File: cmake/generic.cmake (hunk 5 above)
Comment (wopeizl) [reply cmake/generic.cmake]: good catch

File: cmake/external/openblas.cmake [hunk 8]
```diff
@@ -34,66 +30,93 @@ IF(NOT ${CBLAS_FOUND})
...
-        BUILD_IN_SOURCE     1
-        BUILD_COMMAND       ${CMAKE_MAKE_PROGRAM} ${COMMON_ARGS} ${OPTIONAL_ARGS}
-        INSTALL_COMMAND     ${CMAKE_MAKE_PROGRAM} install NO_SHARED=1 NO_LAPACK=1 PREFIX=<INSTALL_DIR> 
-                            && rm -r ${CBLAS_INSTALL_DIR}/lib/cmake ${CBLAS_INSTALL_DIR}/lib/pkgconfig
-        UPDATE_COMMAND      ""
-        CONFIGURE_COMMAND   ""
-    )
-    ELSE()
-    ENDIF(NOT WIN32)
+        IF(WIN32)
```
Comment (reyoung) [inline cmake/external/openblas.cmake]: In windows platform, can we build `openblas` by cmake? If not, why we need this `ExternalProject_Add` ?

File: cmake/external/xxhash.cmake [hunk 9]
```diff
@@ -14,23 +14,52 @@ ELSE()
...
+          -DCMAKE_VERBOSE_MAKEFILE:BOOL=OFF
+          -DCMAKE_POSITION_INDEPENDENT_CODE:BOOL=ON
+          -DBUILD_XXHSUM=OFF
+          -DCMAKE_GENERATOR_PLATFORM=x64
+          -DBUILD_SHARED_LIBS=OFF
+          ${OPTIONAL_CACHE_ARGS}
+          TEST_COMMAND      ""
+  )
+else()
+  ExternalProject_Add(
```
Comment (reyoung) [inline cmake/external/xxhash.cmake:41]: It seems that when linux, we can use cmake for `xxhash`, too. Should we clean these logic?

File: cmake/generic.cmake [hunk 10]
```diff
@@ -289,6 +293,45 @@ function(cc_library TARGET_NAME)
...
+  # the dummy target would be consisted of limit size libraries
+  set(dummy_limit 50)
+  list(LENGTH sep_library_DEPS sep_all_len)
+  foreach(v ${sep_library_DEPS})
+    list(APPEND dummy_list ${v})
+    list(LENGTH dummy_list listlen )
+    if ((${listlen} GREATER ${dummy_limit}) OR (${dummy_offset} EQUAL ${sep_all_len}))
+      message("create dummy library ${TARGET_NAME}_dummy_lib_${dummy_index} for ${TARGET_NAME}")
+      cc_library(${TARGET_NAME}_dummy_lib_${dummy_index} STATIC DEPS ${dummy_list})
+      foreach(i ${dummy_list})
```
Comment (reyoung) [inline cmake/generic.cmake:319]: Maybe we have better way to clean cmake list.

File: cmake/generic.cmake [hunk 11]
```diff
@@ -289,6 +293,45 @@ function(cc_library TARGET_NAME)
...
 
+# The link operation under windows may exceeds the maximum characters limit, simply break the link command
+# into multiple link opeartion can fix that, say
+# original:
+#     lib /out:target.lib a.lib b.lib c.lib d.lib
+# after:
+#    1. lib /out:dummy_lib_1.lib a.lib b.lib
+#    2. lib /out:dummy_lib_2.lib c.lib d.lib
+#    1. lib /out:target.lib dummy_lib_1.lib dummy_lib_2.lib
+function(sep_library TARGET_NAME)
```
Comment (reyoung) [inline cmake/generic.cmake:303]: Perhaps it should be the default behaviour of `cc_library` to split the link command when it is too long.

However, we could polish `cc_library` later.

File: paddle/fluid/framework/CMakeLists.txt [hunk 12]
```diff
@@ -140,16 +140,23 @@ cc_library(proto_desc SRCS var_desc.cc op_desc.cc block_desc.cc program_desc.cc
...
 add_dependencies(framework_py_proto framework_py_proto_init)
-add_custom_command(TARGET framework_py_proto POST_BUILD
-    COMMAND ${CMAKE_COMMAND} -E make_directory ${PADDLE_BINARY_DIR}/python/paddle/fluid/proto
-    COMMAND cp *.py ${PADDLE_BINARY_DIR}/python/paddle/fluid/proto/
-    COMMENT "Copy generated python proto into directory paddle/fluid/proto."
-    WORKING_DIRECTORY ${CMAKE_CURRENT_BINARY_DIR})
+if (NOT WIN32)
+  add_custom_command(TARGET framework_py_proto POST_BUILD
+      COMMAND ${CMAKE_COMMAND} -E make_directory ${PADDLE_BINARY_DIR}/python/paddle/fluid/proto
+      COMMAND cp *.py ${PADDLE_BINARY_DIR}/python/paddle/fluid/proto/
```
Comment (reyoung) [inline paddle/fluid/framework/CMakeLists.txt:155]: Perhaps cmake -E copy can unify this `if/else`.

File: paddle/fluid/framework/details/multi_devices_graph_pass.cc [hunk 13]
```diff
@@ -648,6 +648,12 @@ int MultiDevSSAGraphBuilder::GetVarDeviceID(
     const ir::Graph &graph, const std::string &varname,
     const std::unordered_map<std::string, int> &sharded_var_device) const {
   auto got = sharded_var_device.find(varname);
+  if (got == sharded_var_device.end()) {
```
Comment (reyoung) [inline paddle/fluid/framework/details/multi_devices_graph_pass.cc]: Why we need change this file?

File: paddle/fluid/framework/garbage_collector.h [hunk 14]
```diff
@@ -29,7 +29,7 @@ template <typename T>
 class GarbageCollector {
  public:
   GarbageCollector(const platform::Place &place, size_t max_memory_size)
-      : max_memory_size_(std::max(max_memory_size, static_cast<size_t>(1))) {
+      : max_memory_size_((std::max)(max_memory_size, static_cast<size_t>(1))) {
```
Comment (reyoung) [inline paddle/fluid/framework/garbage_collector.h:32]: Why we need change this file?

File: paddle/fluid/platform/variant.h [hunk 15]
```diff
@@ -42,3 +42,11 @@ limitations under the License. */
 #include <boost/mpl/comparison.hpp>
 #include <boost/mpl/less_equal.hpp>
 #include <boost/variant.hpp>
+
+// some platform-independent defintion
+#if defined(WIN32) || defined(_WIN32) || defined(__WIN32__)
```
Comment (reyoung) [inline paddle/fluid/platform/variant.h]: __attribute__((unused)) is supported by gcc and clang. Not related to platform.

Maybe macro should change to 

```
#if __MSVC__
...
#else
...
#endif
```

**Comment by reyoung** [comment]
Please resolve conflicts.

File: cmake/generic.cmake (hunk 11 above)
Comment (wopeizl) [reply cmake/generic.cmake:303]: ok

File: paddle/fluid/framework/CMakeLists.txt (hunk 12 above)
Comment (wopeizl) [reply paddle/fluid/framework/CMakeLists.txt:155]: ok

File: paddle/fluid/framework/garbage_collector.h (hunk 14 above)
Comment (wopeizl) [reply paddle/fluid/framework/garbage_collector.h:32]: this is the simpliest way to fix a windows build error

File: paddle/fluid/framework/CMakeLists.txt (hunk 12 above)
Comment (wopeizl) [reply paddle/fluid/framework/CMakeLists.txt:155]: seems the copy cannot recognize the *

File: paddle/fluid/framework/details/multi_devices_graph_pass.cc (hunk 13 above)
Comment (wopeizl) [reply paddle/fluid/framework/details/multi_devices_graph_pass.cc]: just merged from develop

File: cmake/external/xxhash.cmake (hunk 9 above)
Comment (wopeizl) [reply cmake/external/xxhash.cmake:41]: ok

File: cmake/external/openblas.cmake [hunk 16]
```diff
@@ -34,66 +30,65 @@ IF(NOT ${CBLAS_FOUND})
...
+            PREFIX              ${CBLAS_SOURCES_DIR}
+            INSTALL_DIR         ${CBLAS_INSTALL_DIR}
+            BUILD_IN_SOURCE     1
+            BUILD_COMMAND       ${CMAKE_MAKE_PROGRAM} ${COMMON_ARGS} ${OPTIONAL_ARGS}
+            INSTALL_COMMAND     ${CMAKE_MAKE_PROGRAM} install NO_SHARED=1 NO_LAPACK=1 PREFIX=<INSTALL_DIR>
+            && rm -r ${CBLAS_INSTALL_DIR}/lib/cmake ${CBLAS_INSTALL_DIR}/lib/pkgconfig
+            UPDATE_COMMAND      ""
+            CONFIGURE_COMMAND   ""
+            )
+    ENDIF (WITH_PREBUILD_OPENBLAS)
```
Comment (dzhwinter) [inline cmake/external/openblas.cmake]: add an option to the whole project will make our options more complex to configure.  Please remove WITH_PREBUILD_OPENBLAS if found the openblas can be built automatically. 
BTW, mkl and mkldnn perform better than openblas, last time I just take a shortcut.

File: paddle/fluid/framework/ir/attention_lstm_fuse_pass.cc [hunk 17]
```diff
@@ -211,12 +211,12 @@ void PrepareLSTMWeight(const LoDTensor& W_forget_w0,
   VLOG(30) << "LSTMWeight resized to " << out->dims();
 
   float* out_data = out->mutable_data<float>(platform::CPUPlace());
-  std::array<const float*, 4> tensors(
-      {{W_forget_w0.data<float>(), W_input_w0.data<float>(),
-        W_output_w0.data<float>(), W_cell_w0.data<float>()}});
-  std::array<const float*, 4> tensors1(
-      {{W_forget_w1.data<float>(), W_input_w1.data<float>(),
-        W_output_w1.data<float>(), W_cell_w1.data<float>()}});
+  std::array<const float*, 4> tensors =
```
Comment (dzhwinter) [inline paddle/fluid/framework/ir/attention_lstm_fuse_pass.cc]: Please confirm with @JiabinYang , is this way can pass the MAC build? I'm not sure about it.

File: paddle/fluid/platform/stream_callback_manager.h [hunk 18]
```diff
@@ -18,8 +18,8 @@
 #include <cuda_runtime.h>
 #include <functional>
 #include <memory>
-#include "ThreadPool.h"
 #include "paddle/fluid/platform/enforce.h"
+#include "third_party/threadpool/src/extern_threadpool/ThreadPool.h"
```
Comment (dzhwinter) [inline paddle/fluid/platform/stream_callback_manager.h]: Please clean up this line. If the include path does not take effect, we need to fix the include cmakelist

File: paddle/fluid/platform/port.h [hunk 19]
```diff
@@ -24,43 +24,42 @@
 #include "glog/logging.h"
 
 #if !defined(_WIN32)
-#define UNUSED __attribute__((unused))
-#include <dlfcn.h>     //  dladdr
-#include <execinfo.h>  // backtrace
-#include <sys/stat.h>
-#include <algorithm>  // std::accumulate
+    #define UNUSED __attribute__((unused))
```
Comment (dzhwinter) [inline paddle/fluid/platform/port.h]: we need to clean up the port header. There is a lot of other UNUSED macro in other files.

**Review by dzhwinter — DISMISSED** [review]
Need to pass the style check.

File: paddle/fluid/platform/port.h (hunk 19 above)
Comment (wopeizl) [reply paddle/fluid/platform/port.h]: ok

File: cmake/external/openblas.cmake (hunk 16 above)
Comment (wopeizl) [reply cmake/external/openblas.cmake]: ok

File: paddle/fluid/platform/stream_callback_manager.h (hunk 18 above)
Comment (wopeizl) [reply paddle/fluid/platform/stream_callback_manager.h]: ok

File: paddle/fluid/framework/ir/attention_lstm_fuse_pass.cc (hunk 17 above)
Comment (wopeizl) [reply paddle/fluid/framework/ir/attention_lstm_fuse_pass.cc]: ok

File: python/CMakeLists.txt [hunk 20]
```diff
@@ -45,23 +45,42 @@ endif()
...
-add_custom_command(OUTPUT ${PADDLE_PYTHON_BUILD_DIR}/.timestamp
-    COMMAND touch stub.cc
-    COMMAND cp -r ${PADDLE_SOURCE_DIR}/python/paddle ${PADDLE_BINARY_DIR}/python
-    COMMAND cp -r ${PADDLE_SOURCE_DIR}/paddle/py_paddle ${PADDLE_BINARY_DIR}/python/
-    COMMAND env ${py_env} ${PYTHON_EXECUTABLE} setup.py bdist_wheel
-    COMMAND ${CMAKE_COMMAND} -E touch ${PADDLE_PYTHON_BUILD_DIR}/.timestamp
-    COMMAND ${CMAKE_COMMAND} -E remove_directory ${PADDLE_PYTHON_BUILD_DIR}/lib-python
-    COMMAND ${CMAKE_COMMAND} -E copy_directory ${PADDLE_PYTHON_BUILD_DIR}/lib.* ${PADDLE_PYTHON_BUILD_DIR}/lib-python
-    DEPENDS gen_proto_py copy_paddle_pybind ${FLUID_CORE} framework_py_proto profiler_py_proto ${PY_FILES} ${external_project_dependencies} ${COPY_PADDLE_MASTER})
+IF(WIN32)
```
Comment (dzhwinter) [inline python/CMakeLists.txt:65]: maybe use cmake -E xxx is fine. No need to split into two commands.

**Review by dzhwinter — APPROVED** [review]
This PR modifies too many files, please polish the cmake script and small fixes after merge.

File: paddle/fluid/framework/ir/pass.h [hunk 21]
```diff
@@ -197,26 +197,26 @@ struct PassRegistrar : public Registrar {
...
+      __reg_pass__##pass_type,                              \
+      "REGISTER_PASS must be called in global namespace");  \
+  static ::paddle::framework::ir::PassRegistrar<pass_class> \
+      __pass_registrar_##pass_type##__(#pass_type);         \
+  int TouchPassRegistrar_##pass_type() {                    \
+    __pass_registrar_##pass_type##__.Touch();               \
+    return 0;                                               \
+  }                                                         \
+  static ::paddle::framework::ir::PassRegistrar<pass_class> \
+      &__pass_tmp_registrar_##pass_type##__ UNUSED =        \
```
Comment (dzhwinter) [inline paddle/fluid/framework/ir/pass.h:211]: Please unify the UNUSED macro with port.h

Comment (dzhwinter) [inline paddle/fluid/framework/ir/pass.h:211]: Please unify the UNUSED macro with port.h

File: python/CMakeLists.txt [hunk 22]
```diff
@@ -45,23 +45,42 @@ endif()
...
+    add_custom_command(OUTPUT ${PADDLE_PYTHON_BUILD_DIR}/.timestamp
+            COMMAND ${CMAKE_COMMAND} -E copy_directory ${PADDLE_SOURCE_DIR}/python/paddle ${PADDLE_BINARY_DIR}/python/paddle/
+            COMMAND ${CMAKE_COMMAND} -E copy_directory ${PADDLE_SOURCE_DIR}/paddle/py_paddle ${PADDLE_BINARY_DIR}/python/
+            COMMAND ${CMAKE_COMMAND} -E env ${py_env} ${PYTHON_EXECUTABLE} setup.py bdist_wheel
+            COMMAND ${CMAKE_COMMAND} -E touch ${PADDLE_PYTHON_BUILD_DIR}/.timestamp
+            COMMAND ${CMAKE_COMMAND} -E remove_directory ${PADDLE_PYTHON_BUILD_DIR}/lib-python
+            DEPENDS gen_proto_py copy_paddle_pybind ${FLUID_CORE} framework_py_proto profiler_py_proto ${PY_FILES} ${external_project_dependencies} ${COPY_PADDLE_MASTER})
+ELSE(WIN32)
+	add_custom_command(OUTPUT ${PADDLE_PYTHON_BUILD_DIR}/.timestamp
+		COMMAND touch stub.cc
```
Comment (dzhwinter) [inline python/CMakeLists.txt:75]: This one can use cmake -E copy_file, maybe we can unify the cmake scripts.

**Comment by wopeizl** [comment]
thanks for great help

File: paddle/fluid/platform/variant.h [hunk 23]
```diff
@@ -42,3 +42,11 @@ limitations under the License. */
 #include <boost/mpl/comparison.hpp>
 #include <boost/mpl/less_equal.hpp>
 #include <boost/variant.hpp>
+
+// some platform-independent defintion
```
Comment (dzhwinter) [inline paddle/fluid/platform/variant.h:46]: And this macro

File: paddle/fluid/pybind/pybind.cc [hunk 24]
```diff
@@ -21,6 +21,13 @@ limitations under the License. */
 #include <utility>
 #include <vector>
 
+#if defined(_WIN32)
+#define NOMINMAX
+#define GLOG_NO_ABBREVIATED_SEVERITIES  // msvc conflict logging with windows.h
+#define GOOGLE_GLOG_DLL_DECL
+#include <Windows.h>
```
Comment (dzhwinter) [inline paddle/fluid/pybind/pybind.cc:28]: use little case <windows.h>, windows has rename the header file.

Comment (dzhwinter) [reply paddle/fluid/pybind/pybind.cc:28]: GLOG_NO_ABBREVIATED_SEVERITIES
GOOGLE_GLOG_DLL_DECL
can be put into a common header.

**Pull request merged**